File: jupyter/Qtjupyter/python/module/backtest.py
import pandas as pd
import numpy as np

from module.broker import Broker
from module.strategy import Strategy

class Backtest:
    def __init__(self,market,strategy, cash=100000):
        self.market=market
        self.broker = Broker(cash=cash,market=market)
        self.strategy = strategy(broker=self.broker,market=market,cash=cash)
    def process_day(self):
        self.broker.processOrders()
        self.strategy.init()
        self.strategy.next()
    def run(self):
        print(" 'Output': 'Running Backtest.'")
        try:
            for i in range(1, self.market.length):
                dailyMarket = self.market.getCutMarket(i)
                self.strategy.market = dailyMarket
                self.broker.market = dailyMarket
                self.process_day()
        except Exception as e:
            print(" 'Output': 'Error Detected.'")
            raise e

        finally:
            print(" 'Output': 'Backtest Complete.'")

File: src/Backtest.jl
#test


export Backtest,
    runTest,
    processDay!

"""

    Backtest(market::Market, strategy::Strategy, cash::Real=1000)

Backtest a strategy on a market.

```jldoctest
Random.seed!(1234);
x=Asset();
y=Asset();
M=Market([x,y]);
T = Backtest(M,CrossOverStrategy,1000)
runTest(T)

# output

Backtest of CrossOverStrategy with 375.71 funds on market comprised of 2 assets.

```
"""
mutable struct Backtest{T<:Strategy}
    market::Market
    broker::Broker
    strategy::T
    completed::Bool
    function Backtest(market::Market, strategy::Type, cash::Real=1000)
        this = new{strategy}()
        this.market = market
        this.broker = Broker(market,cash)
        this.strategy = strategy(this.broker);
        this.completed = false
        return this
    end

end



function processDay!(BT::Backtest)
    processAll!(BT.broker)
    next(BT.strategy)
    return 
end
function runTest(BT::Backtest)
    if BT.completed
        error("Backtest already completed")
    end
    # Initialize strategy
    init(BT.strategy)
    market = cutDataUntil(BT.market,end_date(BT.market))
    BT.broker.market = market
    BT.strategy.market = market 
    for i in getDomain(market)
        takeData!(BT.market,market,i)
       #println(length(BT.broker.market))
        processDay!(BT)
    end
    BT.completed = true
    return BT
end




Base.show(io::IO,BT::Backtest) = print(io,
"""
Backtest of $(typeof(BT.strategy)) with $(round(BT.broker.cash;digits=2)) funds on market comprised of $(length(keys(BT.market.data))) assets.
""")


function plot(BT::Backtest)
    plot(BT.broker)
end





